AcquisitionGeometry prints first and last 10 angles

hrobarts opened the pull request:
## Describe your changes
Updated the information that AcquisitionGeometry prints, to include the first and last 10 angles rather than the first 20

## Describe any testing you have performed
Added test in Test_AcquisitionGeometry 

## Link relevant issues


## Checklist when you are ready to request a review

- [x] I have performed a self-review of my code
- [x] I have added docstrings in line with the guidance in the developer guide
- [x] I have implemented unit tests that cover any new or modified functionality
- [x] CHANGELOG.md has been updated with any functionality change
- [x] Request review from all relevant developers
- [x] Change pull request label to 'Waiting for review'

## Contribution Notes

Please read and adhere to the [developer guide](https://tomographicimaging.github.io/CIL/nightly/developer_guide.html) and local patterns and conventions.

- [x] The content of this Pull Request (the Contribution) is intentionally submitted for inclusion in CIL (the Work) under the terms and conditions of the [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0) License.
- [x] I confirm that the contribution does not violate any intellectual property rights of third parties

lauramurgatroyd commented on Wrappers/Python/test/test_AcquisitionGeometry.py:
```diff
@@ -1609,4 +1609,21 @@ def AcquisitionGeometry_split_to_BlockGeometry(self, data, method, seed):
+            .set_angles(angles=range(90))\
+            .set_labels(['horizontal', 'angle'])
+        AD = AcquisitionData(np.zeros([10,90]), geometry=AG, deep_copy=False)
+        print(AD.geometry)
```
We typically try to avoid printing in the unit tests as it clutters the output.
Maybe you could divert sys,stdout so that despite calling print the output doesn't appear?
Something like this:https://stackoverflow.com/a/34738440

lauramurgatroyd reviewed (changes requested):
It'll be really helpful to see a selection of angles at the start and end 😄 
Just made some suggestions for changes

lauramurgatroyd commented on Wrappers/Python/cil/framework/framework.py:
```diff
@@ -1927,8 +1927,12 @@ def angle_unit(self,val):
+            num_print=max(0,self.num_positions-10)
+            repres += "\tAngles {0}-{1} in {2}s:\n{3}\n".format(num_print, self.num_positions, self.angle_unit, numpy.array2string(self.angle_data[num_print:self.num_positions], separator=', '))
+        return repres 
         return repres   
```
```suggestion
```
The return is duplicated

lauramurgatroyd commented on Wrappers/Python/test/test_AcquisitionGeometry.py:
```diff
@@ -1618,12 +1621,40 @@ def test_geometry_print_angles(self):
         print(AD.geometry)
+        angles = re.findall('Angles [\d]+-[\d]+ in degrees:\s+\[.*\]+', capturedOutput.getvalue(), re.MULTILINE)
+        self.assertEqual(angles[0], 'Angles 0-9 in degrees:\n[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]')
+        self.assertEqual(angles[1], 'Angles 81-90 in degrees:\n[80., 81., 82., 83., 84., 85., 86., 87., 88., 89.]')
```
Sorry I think I said in my previous review that there would be an off by one error in your previous code on this line but seems I was wrong!

lauramurgatroyd replied:
Looks great otherwise!

hrobarts replied:
Good catch, thank you!

gfardell commented:
Can you add a screen shot of the print statement for a dataset with more than 20 angles, with somewhere between 10-20, and with less than 10?

Maybe for a reasonable size (30?) we could print them all, over this (>30?) just print the first and last 10 as you are. It would stop the issue of 21 angles giving you all but one of them.

When the print is truncated we could also concisely tell people how to get all the angles i.e. "Full angular array can be accessed at `acquisition.geometry.angles`" Or `get_angles(units='degree')` could be useful functionality to add.

gfardell reviewed (changes requested):
See comment on PR

hrobarts commented:
> Can you add a screen shot of the print statement for a dataset with more than 20 angles, with somewhere between 10-20, and with less than 10?
> 
> Maybe for a reasonable size (30?) we could print them all, over this (>30?) just print the first and last 10 as you are. It would stop the issue of 21 angles giving you all but one of them.
> 
> When the print is truncated we could also concisely tell people how to get all the angles i.e. "Full angular array can be accessed at `acquisition.geometry.angles`" Or `get_angles(units='degree')` could be useful functionality to add.

That makes sense - I've updated the code so it's now printing all angles if there are up to 30, but otherwise printing first and last 10. I've also added the acquisition.geometry.angles suggestion

gfardell reviewed (approved):
Add to the changelog, but approved

hrobarts commented on Wrappers/Python/cil/framework/framework.py:
```diff
@@ -1927,8 +1927,12 @@ def angle_unit(self,val):
+        if num_print < self.num_positions:
+            num_print=max(0,self.num_positions-10)
+            repres += "\tAngles {0}-{1} in {2}s:\n{3}\n".format(num_print, self.num_positions, self.angle_unit, numpy.array2string(self.angle_data[num_print:self.num_positions], separator=', '))
+        return repres 
```
Thanks Laura, this is updated now

lauramurgatroyd commented on Wrappers/Python/cil/framework/framework.py:
```diff
@@ -1927,9 +1927,14 @@ def angle_unit(self,val):
-        return repres   
+        # max_num_print = 30 
+        if self.num_positions < 31:
+            repres += "\tAngles 0-{0} in {1}s: {2}\n".format(self.num_positions, self.angle_unit, numpy.array2string(self.angle_data[0:self.num_positions], separator=', '))
```
```suggestion
            repres += "\tAngles 0-{0} in {1}s: {2}\n".format(self.num_positions-1, self.angle_unit, numpy.array2string(self.angle_data[0:self.num_positions], separator=', '))
```

lauramurgatroyd replied:
Sorry, another change but say we have num_positions = 30, it should say angles 0-29

lauramurgatroyd replied:
Also I don't think there was a unit test for this case

lauramurgatroyd reviewed (changes requested):
Sorry I think I spotted another -1 issue 😢

hrobarts commented:
For angles = 30
![image](https://github.com/TomographicImaging/CIL/assets/77114597/4f39ac60-5c2b-4d2a-933b-b83f936f2404)

For angles = 31
![image](https://github.com/TomographicImaging/CIL/assets/77114597/34e0785b-8c48-4f89-895c-633bb7068341)

hrobarts commented:
> Sorry I think I spotted another -1 issue 😢

Thanks Laura - it's updated now, and I've added a test

lauramurgatroyd reviewed (approved):
🥳
